### PR TITLE
pam_unix: allow empty passwords with non-empty hashes

### PR DESCRIPTION
**Pull request by trofi**
Before the change pam_unix has different behaviours for a user with empty password for these two `/etc/shadow` entries:

    nulloktest:$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0:1::::::
    nulloktest::1::::::

The entry with a hash was rejected and the entry without was accepted.

The rejection happened because 9e74e90147c "pam_unix: avoid determining if user exists" introduced the following rejection check (slightly simplified):

        ...
        } else if (p[0] == '\0' && nullok) {
                if (hash[0] != '\0') {
                        retval = PAM_AUTH_ERR;
                }

We should not reject the user with a hash assuming it's non-empty. The change does that by pushing empty password check into `verify_pwd_hash()`.

`NixOS` generates such hashed entries for empty passwords as if they were non-empty using the following perl code:

    sub hashPassword {
        my ($password) = @_;
        my $salt = "";
        my @chars = ('.', '/', 0..9, 'A'..'Z', 'a'..'z');
        $salt .= $chars[rand 64] for (1..8);
        return crypt($password, '$6$' . $salt . '$');
    }

Resolves: https://github.com/linux-pam/linux-pam/issues/758
Fixes: 9e74e90147c "pam_unix: avoid determining if user exists"

### CODE REVIEW

File: modules/pam_unix/passverify.c [hunk 1]
```diff
@@ -76,9 +76,15 @@ PAMH_ARG_DECL(int verify_pwd_hash,
 
 	strip_hpux_aging(hash);
 	hash_len = strlen(hash);
+
+	/* The passed password is empty */
+	if (p && p[0] == '\0' && !nullok) {
+		retval = PAM_AUTH_ERR;
+	}
```
Comment (ldv-alt) [inline modules/pam_unix/passverify.c:83]: Hmm, this assignment doesn't seem to have any effect as `retval` is getting assigned later in this function anyway.

Comment (trofi) [reply modules/pam_unix/passverify.c]: Great catch! Forgot to test the `nullok == 0` case and missed it. Updated it to reside in `if {} else {}` branches.

Tested the cases with:

```c
               // empty request and entry, correct password:
               assert(verify_pwd_hash("", "", 1) == PAM_SUCCESS);
               assert(verify_pwd_hash("", "", 0) != PAM_SUCCESS);
               assert(verify_pwd_hash("", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 1) == PAM_SUCCESS);
               // this case:
               assert(verify_pwd_hash("", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 0) != PAM_SUCCESS);

               // empty request and entry, incorrect password:
               assert(verify_pwd_hash("1", "", 1) != PAM_SUCCESS);
               assert(verify_pwd_hash("1", "", 0) != PAM_SUCCESS);
               assert(verify_pwd_hash("1", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 1) != PAM_SUCCESS);
               assert(verify_pwd_hash("1", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 0) != PAM_SUCCESS);

               // non-empty, correct password:
               assert(verify_pwd_hash("test", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 1) == PAM_SUCCESS);
               assert(verify_pwd_hash("test", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 0) == PAM_SUCCESS);
               // non-empty, incorrect password:
               assert(verify_pwd_hash("test2", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 1) != PAM_SUCCESS);
               assert(verify_pwd_hash("test2", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 0) != PAM_SUCCESS);
```

The case 4 was broken, works now.

**Review by ldv-alt — APPROVED** [review]

**Comment by ldv-alt** [comment]
> @trofi @ldv-alt I noticed that this change causes a delay for login(1) right after entering the user name.

If this happens, it's a regression from 9e74e90147c530801e3ea3428d64371722c90e01 that was made to address information leaks.

**Comment by trofi** [comment]
> @trofi @ldv-alt I noticed that this change causes a delay for login(1) right after entering the user name. So with this change, we now essentially leak information that a specific user exists on the machine: a nonexistent user will continue to the password prompt immediately, whereas any user that has a password hash set will cause a delay while we check against the empty password.

I was not able to reproduce it locally. Do you use any special settings in `/etc/pam.d/login`?

Mine are:

```
# $ cat /etc/pam.d/login |& unnix
cat /etc/pam.d/login | unnix
# Account management.
account required pam_unix.so # unix (order 10900)

# Authentication management.
auth sufficient pam_unix.so likeauth nullok try_first_pass # unix (order 11500)
auth required pam_deny.so # deny (order 12300)

# Password management.
password sufficient pam_unix.so nullok yescrypt # unix (order 10200)

# Session management.
session required pam_env.so conffile=/etc/pam/environment readenv=0 # env (order 10100)
session required pam_unix.so # unix (order 10200)
session required pam_loginuid.so # loginuid (order 10300)
session required /<<NIX>>/linux-pam-1.6.1/lib/security/pam_lastlog.so silent # lastlog (order 10700)
session optional /<<NIX>>/systemd-255.4/lib/security/pam_systemd.so # systemd (order 12000)
```

How I tested:

```
$ sudo su -
# login

nz login: doesnotexist
Password:
(instant "Password:", delayed result on invalid password, that user does not exist)

nz login: root
Password:
(instant "Password:", delayed result on invalid password, that user exists, allwoed password)
```

I see the same results on both `pam-1.6.1` and `pam-1.6.0` with 8d0c575336ad301cd14e16ad2fdec6fe621764b8 applied on top.

**Comment by pks-t** [comment]
On Tue, Apr 30, 2024 at 03:07:36PM -0700, Sergei Trofimovich wrote:
> > @trofi @ldv-alt I noticed that this change causes a delay for
> > login(1) right after entering the user name. So with this change, we
> > now essentially leak information that a specific user exists on the
> > machine: a nonexistent user will continue to the password prompt
> > immediately, whereas any user that has a password hash set will
> > cause a delay while we check against the empty password.
> 
> I was not able to reproduce it locally. Do you use any special settings in `/etc/pam.d/login`?

No, I have the exact same configuration as you do on nixos-unstable.
I pinned down the regression into the range between

    66adc1e47f87 (Merge pull request #304601 from andresilva/firefox-125.0.1, 2024-04-16)

and

    5c24cf2f0a12 (Merge pull request #303600 from r-ryantm/auto-update/python311Packages.mailchecker, 2024-04-19)

This range contains e3201828048c (Merge pull request #298994 from
trofi/linux-pam-revert-helper, 2024-03-29), which backports this pull
request here.

I didn't pin it down exactly to that commit because it causes almost a
world rebuild of my system when I go further into that range. In any
case, the commit range does not contain any other changes that would be
related to authentication, as far as I can see.

> How I tested:
> 
> ```
> $ sudo su -
> # login
> 
> nz login: doesnotexist
> Password:
> (instant "Password:", delayed result on invalid password, that user does not exist)
> 
> nz login: root
> Password:
> (instant "Password:", delayed result on invalid password, that user exists, allwoed password)
> ```

Yeah, this is how I can reproduce the issue. What may be of interest is
that I use yescrypt with a weight of 11 (mkpasswd -m yescrypt -R 11).
This does take a few seconds to decrypt and thus makes the delay after
entering the username quite noticeable.

**Comment by trofi** [comment]
Aha, that makes sense. Reproduced locally. Another (slightly more user visible test) is to set an empty `""` hashed password on a user. I think `login` is expected to ask for a password prompt here, but it does not and logs the user in after a `Login:` prompt. Will double-check against `pam-1.5.3`. Looking.

**Comment by dchmelik** [comment]
Neither should be rejected; they're not on FreeBSD UNIX (a genetic UNIX) and doing so is non-UNIX-like behaviour forcing users & system administrators (sysadmins) to use things the way they might dislike, including computers in secure areas or without important data, and with chroots & virtual machines on which password is simply overkill.  Should I open a new issue about this or would you consider classic UNIX (originally in laboratories people could use without password, and even some users did) standards and update the issue?
